src/nvim/event/loop.c

```diff
@@ -138,6 +138,39 @@ static void loop_walk_cb(uv_handle_t *handle, void *arg)
 }
 #endif
 
+static bool loop_uv_close(uv_loop_t *loop, bool wait)
+{
+  // Run the loop to tickle close-callbacks (which should then free memory).
+  // Use UV_RUN_NOWAIT to avoid a hang. #11820
+  uv_run(loop, UV_RUN_NOWAIT);
```

Comment:
Now the code is written this way, I'm not sure what this line is really doing. How is it tickling the close callbacks?

@justinmk

Reply:
Why specifically would it tickle close callbacks?

Reply:
> Did you mean just to run the loop once to execute any pending callbacks which _may_ be close callbacks? 

yep

> If so, why wouldn't the `uv_run` call below also do this?

It would, but that may not be reached, depending on `wait`.

Would need to study the git logs to be extra sure. This logic is very tricky, driven by various bug reports.

Reply:
> For context luv doesn't do this so just trying to understand the differences. For reference luv just does:

luv doesn't care about hangs, whereas we need to exit quickly.

Reply:
> It would, but that may not be reached, depending on wait.

Oh so we can simplify this further then with something like?

```c
  if (!wait) {
    // Run the loop to run any pending callbacks (which may then free memory).
    // Use UV_RUN_NOWAIT to avoid a hang. #11820
    uv_run(loop, UV_RUN_NOWAIT);
    uv_loop_close(loop);
    return true;
  }
```

> luv doesn't care about hangs, whereas we need to exit quickly.

Shouldn't it? If our code is _better_ than theirs, then I'd be happy to port some of it over. We use their loop handling for our Lua threads.

Reply:
> If our code is _better_ than theirs, then I'd be happy to port some of it over. We use their loop handling for our Lua threads.

It's arguably a logic bug in our system. In theory we should never "hang", but somehow there are (or were) cases where we run `UV_RUN_DEFAULT` and there's nothing on the loop, so we hang forever.

In other cases, on exit, users simply don't want to wait for plugins/jobs (slow child processes) to cleanup.